File: lib/xairo/context.ex
defmodule Xairo.Context do
  @moduledoc false

  defstruct [:context]

  def new(%Xairo.ImageSurface{surface: surface}) do
    with {:ok, context} <- Xairo.Native.context_new_from_image_surface(surface) do
      %__MODULE__{
        context: context
      }
    end
  end

  def status(%__MODULE__{context: ctx}) do
    with {:ok, _} <- Xairo.Native.context_status(ctx), do: :ok
  end

  def set_source_rgb(%__MODULE__{context: ctx} = this, r, g, b) do
    Xairo.Native.context_set_source_rgb(ctx, r / 1, g / 1, b / 1)
    this
  end

  def set_source_rgba(%__MODULE__{context: ctx} = this, r, g, b, a) do
    Xairo.Native.context_set_source_rgba(ctx, r / 1, g / 1, b / 1, a / 1)
    this
  end

  def paint(%__MODULE__{context: ctx} = this) do
    with {:ok, _} <- Xairo.Native.context_paint(ctx), do: this
  end

  def paint_with_alpha(%__MODULE__{context: ctx} = this, alpha) do
    with {:ok, _} <- Xairo.Native.context_paint_with_alpha(ctx, alpha / 1), do: this
  end

  def stroke(%__MODULE__{context: ctx} = this) do
    with {:ok, _} <- Xairo.Native.context_stroke(ctx), do: this
  end

  def stroke_preserve(%__MODULE__{context: ctx} = this) do
    with {:ok, _} <- Xairo.Native.context_stroke_preserve(ctx), do: this
  end

  def fill(%__MODULE__{context: ctx} = this) do
    with {:ok, _} <- Xairo.Native.context_fill(ctx), do: this
  end

  def fill_preserve(%__MODULE__{context: ctx} = this) do
    with {:ok, _} <- Xairo.Native.context_fill_preserve(ctx), do: this
  end

  def move_to(%__MODULE__{context: ctx} = this, x, y) do
    Xairo.Native.context_move_to(ctx, x / 1, y / 1)
    this
  end

  def line_to(%__MODULE__{context: ctx} = this, x, y) do
    Xairo.Native.context_line_to(ctx, x / 1, y / 1)
    this
  end

  def curve_to(%__MODULE__{context: ctx} = this, x1, y1, x2, y2, x3, y3) do
    Xairo.Native.context_curve_to(ctx, x1 / 1, y1 / 1, x2 / 1, y2 / 1, x3 / 1, y3 / 1)
    this
  end

  def rel_move_to(%__MODULE__{context: ctx} = this, x, y) do
    Xairo.Native.context_rel_move_to(ctx, x / 1, y / 1)
    this
  end

  def rel_line_to(%__MODULE__{context: ctx} = this, x, y) do
    Xairo.Native.context_rel_line_to(ctx, x / 1, y / 1)
    this
  end

  def rel_curve_to(%__MODULE__{context: ctx} = this, x1, y1, x2, y2, x3, y3) do
    Xairo.Native.context_rel_curve_to(ctx, x1 / 1, y1 / 1, x2 / 1, y2 / 1, x3 / 1, y3 / 1)
    this
  end

  def arc(%__MODULE__{context: ctx} = this, cx, cy, r, a1, a2) do
    Xairo.Native.context_arc(ctx, cx / 1, cy / 1, r / 1, a1 / 1, a2 / 1)
    this
  end

  def arc_negative(%__MODULE__{context: ctx} = this, cx, cy, r, a1, a2) do
    Xairo.Native.context_arc_negative(ctx, cx / 1, cy / 1, r / 1, a1 / 1, a2 / 1)
    this
  end

  def new_path(%__MODULE__{context: ctx} = this) do
    Xairo.Native.context_new_path(ctx)
    this
  end

  def new_sub_path(%__MODULE__{context: ctx} = this) do
    Xairo.Native.context_new_sub_path(ctx)
    this
  end

  def rectangle(%__MODULE__{context: ctx} = this, x, y, w, h) do
    Xairo.Native.context_rectangle(ctx, x / 1, y / 1, w / 1, h / 1)
    this
  end

  def close_path(%__MODULE__{context: ctx} = this) do
    Xairo.Native.context_close_path(ctx)
    this
  end
end

File: native/xairo/src/context.rs
use crate::error::XairoError;
use crate::image_surface::ImageSurface;
use rustler::ResourceArc;

pub struct ContextRaw {
    pub context: cairo::Context,
}

unsafe impl Send for ContextRaw {}
unsafe impl Sync for ContextRaw {}

pub type Context = ResourceArc<ContextRaw>;

#[rustler::nif]
fn context_new_from_image_surface(surface: ImageSurface) -> Result<Context, XairoError> {
    match cairo::Context::new(&surface.surface) {
        Ok(context) => Ok(ResourceArc::new(ContextRaw { context })),
        Err(err) => Err(err.into()),
    }
}

#[rustler::nif]
fn context_status(context: Context) -> Result<(), XairoError> {
    match context.context.status() {
        Ok(_) => Ok(()),
        Err(err) => Err(err.into()),
    }
}

#[rustler::nif]
fn context_set_source_rgb(context: Context, r: f64, g: f64, b: f64) {
    context.context.set_source_rgb(r, g, b);
}

#[rustler::nif]
fn context_set_source_rgba(context: Context, r: f64, g: f64, b: f64, a: f64) {
    context.context.set_source_rgba(r, g, b, a);
}

#[rustler::nif]
fn context_paint(context: Context) -> Result<(), XairoError> {
    match context.context.paint() {
        Ok(_) => Ok(()),
        Err(err) => Err(err.into()),
    }
}

#[rustler::nif]
fn context_paint_with_alpha(context: Context, alpha: f64) -> Result<(), XairoError> {
    match context.context.paint_with_alpha(alpha) {
        Ok(_) => Ok(()),
        Err(err) => Err(err.into()),
    }
}

#[rustler::nif]
fn context_fill(context: Context) -> Result<(), XairoError> {
    match context.context.fill() {
        Ok(_) => Ok(()),
        Err(err) => Err(err.into()),
    }
}

#[rustler::nif]
fn context_fill_preserve(context: Context) -> Result<(), XairoError> {
    match context.context.fill_preserve() {
        Ok(_) => Ok(()),
        Err(err) => Err(err.into()),
    }
}

#[rustler::nif]
fn context_stroke(context: Context) -> Result<(), XairoError> {
    match context.context.stroke() {
        Ok(_) => Ok(()),
        Err(err) => Err(err.into()),
    }
}

#[rustler::nif]
fn context_stroke_preserve(context: Context) -> Result<(), XairoError> {
    match context.context.stroke_preserve() {
        Ok(_) => Ok(()),
        Err(err) => Err(err.into()),
    }
}

#[rustler::nif]
fn context_move_to(context: Context, x: f64, y: f64) {
    context.context.move_to(x, y);
}

#[rustler::nif]
fn context_rel_move_to(context: Context, x: f64, y: f64) {
    context.context.rel_move_to(x, y);
}

#[rustler::nif]
fn context_line_to(context: Context, x: f64, y: f64) {
    context.context.line_to(x, y);
}

#[rustler::nif]
fn context_rel_line_to(context: Context, x: f64, y: f64) {
    context.context.rel_line_to(x, y);
}

#[rustler::nif]
fn context_rectangle(context: Context, x: f64, y: f64, w: f64, h: f64) {
    context.context.rectangle(x, y, w, h);
}

#[rustler::nif]
fn context_close_path(context: Context) {
    context.context.close_path();
}

#[rustler::nif]
fn context_curve_to(context: Context, x1: f64, y1: f64, x2: f64, y2: f64, x3: f64, y3: f64) {
    context.context.curve_to(x1, y1, x2, y2, x3, y3);
}

#[rustler::nif]
fn context_rel_curve_to(context: Context, x1: f64, y1: f64, x2: f64, y2: f64, x3: f64, y3: f64) {
    context.context.rel_curve_to(x1, y1, x2, y2, x3, y3);
}

#[rustler::nif]
fn context_arc(context: Context, cx: f64, cy: f64, r: f64, a1: f64, a2: f64) {
    context.context.arc(cx, cy, r, a1, a2);
}

#[rustler::nif]
fn context_arc_negative(context: Context, cx: f64, cy: f64, r: f64, a1: f64, a2: f64) {
    context.context.arc_negative(cx, cy, r, a1, a2);
}

#[rustler::nif]
fn context_new_path(context: Context) {
    context.context.new_path();
}

#[rustler::nif]
fn context_new_sub_path(context: Context) {
    context.context.new_sub_path();
}
